frontend/services/gui/src/windows/HotkeyEditor.cpp

```diff
@@ -0,0 +1,313 @@
+
+    return true;
+}
+
+
+void megamol::gui::HotkeyEditor::SpecificStateFromJSON(const nlohmann::json& in_json) {
+
+    if (this->command_registry_ptr != nullptr) {
+        for (auto& header_item : in_json.items()) {
+            if (header_item.key() == GUI_JSON_TAG_WINDOW_CONFIGS) {
+                for (auto& window_item : header_item.value().items()) {
+                    if (window_item.key() == this->Name()) {
+                        for (auto& config_item : window_item.value().items()) {
+                            if (config_item.key() == "hotkey_list") {
+                                /// TODO Add verbose log on error?
+                                if (config_item.value().is_array()) {
+                                    for (auto& cmds_array : config_item.value().items()) {
+                                        if (cmds_array.value().is_array()) {
+                                            for (auto& cmd_json : cmds_array.value().items()) {
+                                                megamol::frontend_resources::Command cmd;
+                                                megamol::frontend_resources::from_json(cmd_json.value(), cmd);
+                                                if (!this->command_registry_ptr->update_hotkey(cmd.name, cmd.key)) {
+                                                    switch (cmd.parent_type) {
+                                                    case (megamol::frontend_resources::Command::parent_type_c::
+                                                            PARENT_PARAM):
+                                                        cmd.effect = this->megamolgraph_ptr->Parameter_Lambda;
+                                                        break;
+                                                    case (megamol::frontend_resources::Command::parent_type_c::
+                                                            PARENT_GUI_HOTKEY):
+                                                        cmd.effect = this->parent_gui_hotkey_lambda;
+                                                        break;
+                                                    case (megamol::frontend_resources::Command::parent_type_c::
+                                                            PARENT_GUI_WINDOW):
+                                                        cmd.effect = this->parent_gui_window_lambda;
+                                                        break;
+                                                    case (megamol::frontend_resources::Command::parent_type_c::
+                                                            PARENT_GUI_WINDOW_HOTKEY):
+                                                        cmd.effect = this->parent_gui_window_hotkey_lambda;
+                                                        break;
+                                                    }
+                                                    this->command_registry_ptr->add_command(cmd);
```

Comment:
I am not sure why the GUI Hotkey window needs to serialize and reload the whole command state. If Commands come from a certain graph/GUI state, wouldnt the Command Registry be repopulated with the same commands when loading the project file? Doesnt it suffice to just save the (maybe non-default) key<->command bindings instead of the whole command registry?

Reply:
we could add tracking of non-default keys. but the problem is: if modules are serialized in a different order than they were created in the original project file, other commands will get the overlapping hotkey (first come, first serve) - like when you have multiple view::resetView commands. then you need to have all the hotkey information anyway.